pkg/agent/wireguard/client_linux.go

```diff
@@ -0,0 +1,256 @@
+// +build linux
+
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wireguard
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"sync"
+
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"golang.zx2c4.com/wireguard/wgctrl"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/types"
+)
+
+const (
+	ifName            = "antrea-wg0"
+	routeRulePriority = 10
+	routeTable        = 10
+)
+
+var zeroKey = wgtypes.Key{}
+
+// wgctrlClient is an interface to mock wgctrl.Client
+type wgctrlClient interface {
+	io.Closer
+	Devices() ([]*wgtypes.Device, error)
+	Device(name string) (*wgtypes.Device, error)
+	ConfigureDevice(name string, config wgtypes.Config) error
+}
+
+var _ Interface = (*client)(nil)
+
+type client struct {
+	wgClient                wgctrlClient
+	nodeName                string
+	k8sClient               clientset.Interface
+	privateKey              wgtypes.Key
+	peerPublicKeyByNodeName *sync.Map
+	nodeIP                  net.IP
+	wireGuardConfig         *config.WireGuardConfig
+}
+
+func New(clientSet kubernetes.Interface, nodeConfig *config.NodeConfig, wireGuardConfig *config.WireGuardConfig) (Interface, error) {
+	wgClient, err := wgctrl.New()
+	if err != nil {
+		return nil, err
+	}
+	c := &client{
+		wgClient:                wgClient,
+		nodeName:                nodeConfig.Name,
+		k8sClient:               clientSet,
+		nodeIP:                  nodeConfig.NodeIPAddr.IP,
+		wireGuardConfig:         wireGuardConfig,
+		peerPublicKeyByNodeName: &sync.Map{},
+	}
+	wireGuardConfig.WireGuardAvailable = c.PeerConfigured
+	return c, nil
+}
+
+func (client *client) Init() error {
+	link := &netlink.Wireguard{LinkAttrs: netlink.LinkAttrs{Name: ifName, MTU: client.wireGuardConfig.MTU}}
+	err := netlink.LinkAdd(link)
+	// ignore existing link as it may have already been created or managed by userspace process.
+	if err != nil && !errors.Is(err, unix.EEXIST) {
+		if errors.Is(err, unix.EOPNOTSUPP) {
+			return fmt.Errorf("WireGuard not supported by the Linux kernel (netlink: %w)", err)
```

Comment:
Maybe you can add "Make sure wireguard kernel module is loaded".

Reply:
Do you mean by adding the message to the error or the comment?

Reply:
The error log message.